cms_security/models/security.py

```diff
@@ -0,0 +1,161 @@
+    """Provide basic logic for protecting website items.
+
+    TODO: explain in detail what we do.
+    """
+
+    _name = "cms.security.mixin"
+    _description = "A mixin for protecting website content"
+    # internal flag to enable/disable auto creation of ACL + security rules
+    _cms_auto_security_policy = False
+
+    read_group_ids = fields.Many2many(
+        string='View Groups',
+        comodel_name='res.groups',
+        help=("Allow `view` access to this item to specific groups. "
+              u"No group means anybody with proper rights can see it.")
+    )
+    write_group_ids = fields.Many2many(
+        string='Write Groups',
+        comodel_name='res.groups',
+        help=("Allow `write` access to this item to specific groups. "
+              u"No group means anybody with proper rights can edit it.")
+    )
+
+    def _auto_rule_name_prefix(self):
+        """Return name prefix for automatic rules."""
+        return 'cms_security_{}'.format(self._name.replace('.', '_'))
+
+    def _auto_get_model_id(self):
+        """Return ID for current model for automatic rules."""
+        self.env.cr.execute(
+            "SELECT id FROM ir_model WHERE model = %s", (self._name, ))
+        res = self.env.cr.fetchone()
+        return res[0] if res else None
+
+    def _auto_access_values(self, env, model_id):
+        """Return automatic ACL rules values."""
+        prefix = self._auto_rule_name_prefix()
+        return utils.build_ACL_values(env, prefix, model_id)
+
+    def _auto_create_access(self, model_id):
+        """Create automatic ACL."""
+        env = self.env
+        for values in self._auto_access_values(env, model_id):
+            xmlid = values.pop('_xmlid')
+            if not env.ref(xmlid, raise_if_not_found=0):
+                record = env['ir.model.access'].create(values)
+                utils.add_xmlid(env, record, xmlid)
+                logger.info('Created default ACL for %s: %s',
+                            self._name, values['name'])
+
+    def _auto_rule_values(self, env, model_id):
+        """Return automatic record rules values."""
+        prefix = self._auto_rule_name_prefix()
+        return utils.build_rule_values(env, prefix, model_id)
+
+    def _auto_create_rule(self, model_id):
+        """Create automatic record rules."""
+        env = self.env
+        for values in self._auto_rule_values(env, model_id):
+            xmlid = values.pop('_xmlid')
+            if not env.ref(xmlid, raise_if_not_found=0):
+                record = env['ir.rule'].create(values)
+                utils.add_xmlid(env, record, xmlid)
+                logger.info('Created default record rule for %s: %s',
+                            self._name, values['name'])
+
+    @api.model
+    def _setup_complete(self):     # pylint: disable=R8110
+        """Override to handle automatic security."""
+        super()._setup_complete()
+        # TODO: this code runs 2 times on every reboot.
```

Comment:
@sbidoul @guewen can you give me your insights about this?

Probably a possible solution would be to add an API method to be called by on pre/post hooks on modules installation (modules that rely on this feature) but I'd like to keep this automatic and setup everything when a new model is registered.

Reply:
It seems to me that `_setup_*` methods normally "initialize" the state of the models but don't have any side effect on the database. That could explain why it's called 2 times. Modifications in database are done in `_auto_init` and `init` called at addons upgrade / install. 

```python
    def init(self):
        """ This method is called after :meth:`~._auto_init`, and may be
            overridden to create or modify a model's database schema.
        """
        pass
```

What would be missing with an override of `init`?

Reply:
> but I'd like to keep this automatic and setup everything when a new model is registered.

Like when you create a new model from the UI?

Reply:
not just from the UI,  I'd like to catch both UI and code setup

Reply:
@guewen I've just give a run w/ `init`: same result. It's called twice and apparently there's no way to state that everything is already in place. Seems like the only good candidate is `ir.model` create, which very likely should be called just once.